### DISCUSSION TITLE
Scaled Pearson residual

### DISCUSSION THREAD

**Discussion by Vu1992** [casact/chainladder-python#508] [General]
Hi,

In the ODP bootstrap method, i see that we need to calculate (Scaled) Pearson Residual and bootstrap it instead of bootstrap the origin triangle. Can someone show me how to calculate it by this package?

**Comment by kennethshsu** [comment]
Hi @Vu1992, did you get a chance to follow the [tutorial](https://chainladder-python.readthedocs.io/en/latest/getting_started/tutorials/stochastic-tutorial.html#odp-bootstrap-model) or review the [API docs](https://chainladder-python.readthedocs.io/en/latest/library/generated/chainladder.BootstrapODPSample.html)?

I'm not sure what you are trying to do, but it's often not necessary to perform intermediary steps when using an IBNR model.

Please take a look at the links I reference above and let me know if they are helpful or not.

**Reply by Vu1992** [reply]
Hi Kenne,
Thank for your support. Actually, I want to manually adjust the Scaled pearson residual but the formular only let me drop it out. Is there any way you could help me with that?
Also, for the ODP boostrap, after create the false history bootstrap triangle, we need to simulate the incremental payment from Gamma Distribution . Can you show me how we can do that by this package?
Thanks a lot

**Reply by kennethshsu** [reply]
Hi @Vu1992,

I don't think the package allows for manual adjustment now. It will only simulate and return the sampled triangles back to the user for further analytics.

I don't quite follow what you mean by this:
> after create the false history bootstrap triangle, we need to simulate the incremental payment from Gamma Distribution

Can you elaborate more on what you are trying to accomplish? Perhaps, you can share your sample code that you have?

**Reply by Vu1992** [reply]
Hi @kennethshsu ,

I think it will be better for actuary to use this package in practice if the ODP boostrap sample is seperated by two steps, select adjusted scale pearson residual and then bootstrap to get the fallen history triangle.

After we got the ODP bootstrap triangle sample and square it (let call it step 3), we will got the estimation error but not the process variance. In order to get the process variance, we will simulate the incremental payment from a series of gamma distribution with mean equal the projected payment in step3 and variance equal projected payment in step 3 * scale used in calculation scaled pearson residual, then apply the sign of random payment be the same of projected payment.
For more refferrence, u can see the file i attached, the Appendix II: Step by step of the boostrap process for reserve simulation.
[Boostraping reserve.pdf](https://github.com/casact/chainladder-python/files/14925868/Boostraping.reserve.pdf)

I think the main point here that after got the ODP bootstrap sample, we only can get the full triangle square by cl.chainladder but not get the process variance i mentioned above. I think chainladder in R package have done it by the following code "B <- BootChainLadder(RAA, R=999, process.distr="gamma")"

**Reply by jbogaardt** [reply]
Late to the party - each of the IBNR methods (chainladder, benktander, etc.) invokes the [_get_process_variance](https://github.com/casact/chainladder-python/blob/eb8b303a950aff1491ee11e4ece147aecd95077c/chainladder/methods/base.py#L143-L151) method which checks whether the estimator is operating on a bootstrapped triangle. This should invoke the same function in the [bootstrap.py](https://github.com/casact/chainladder-python/blob/eb8b303a950aff1491ee11e4ece147aecd95077c/chainladder/adjustments/bootstrap.py#L256-L268) module where the gamma sampling occurs. So I think @Vu1992  is correct that this is a two step process with the second step being invoked in the IBNR methods.

I think this is what you're looking for - though there is no manual selection step exposed as written.  Happy to take suggestions on how it might be improved.

**Reply by kennethshsu** [reply]
Thank you for stepping in and helping out @jbogaardt! Learning something new everyday!

**Reply by Vu1992** [reply]
Hi @jbogaardt, 
Thank for pointing out the gamma sampling. It mean that when i use IBNR method such as cl.chainladder, it will have the gamma sampling for projected incremental, right?
If it is the case, i think the only suggestion i have is somehow the package allow for manual adjustment to scaled pearson residual because actuaries will do that a lot in practice.

**Reply by jbogaardt** [reply]
Do you envision something simple as a `scale` hyperparameter as a direct input in `BootstrapODPSample` that would be used (if exists) in lieu of calculating from the data?

**Reply by Vu1992** [reply]
As i understand that scaled hyperparamerter is used for all residuals, I think it will be hard for getting an input like that cause adjustment is usually made for some residuals only, and other will still relied on the calculation from data. I see the package got the drop function in the boostrap, maybe can add something similar like [AY, development,replace residual value]

**Reply by Vu1992** [reply]
Sorry one more point is that you need to see the Scaled residual triangle to make any adjustment if needed. So that we will need a attribute to show out the scaled pearson residual triangle.
